raw-spans-grouper/raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/RawSpansProcessor.java

```diff
@@ -196,7 +240,97 @@ public void process(Record<TraceIdentity, RawSpan> record) {
         .record(Duration.between(start, Instant.now()).toMillis(), TimeUnit.MILLISECONDS);
     // no need to do context.forward. the punctuator will emit the trace once it's eligible to be
     // emitted
-    return;
+  }
+
+  private void processSpanForPeerServiceNameIdentification(
+      TraceIdentity key, RawSpan value, TraceState traceState, long currentTimeMs) {
+    Event event = value.getEvent();
+    String tenantId = key.getTenantId();
+    ByteBuffer traceId = value.getTraceId();
+    boolean firstEntry = (traceState == null);
+    final Optional<String> maybeEnvironment =
+        HttpSemanticConventionUtils.getEnvironmentForSpan(event);
+    if (SpanSemanticConventionUtils.isClientSpanForOCFormat(
+        event.getAttributes().getAttributeMap())) {
+      final Optional<String> maybeHostAddr = HttpSemanticConventionUtils.getHostIpAddress(event);
+      final Optional<String> maybePeerAddr = HttpSemanticConventionUtils.getPeerIpAddress(event);
+      final Optional<String> maybePeerPort = HttpSemanticConventionUtils.getPeerPort(event);
+      final String serviceName = event.getServiceName();
+      final IpIdentity ipIdentity =
+          IpIdentity.newBuilder()
+              .setTenantId(tenantId)
+              .setEnvironment(maybeEnvironment.orElse(null))
+              .setHostAddr(maybeHostAddr.orElse(null))
+              .setPeerAddr(maybePeerAddr.orElse(null))
+              .setPeerPort(maybePeerPort.orElse(null))
+              .build();
+      if (IpIdentityValidator.isValid(ipIdentity)) {
+        ipsToSpanMetadataStateStore.put(
```

Comment:
how are we cleaning this up? we need to refresh this information else this will provide wrong results